### PR TITLE
Python Regressions

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

This test changes Python Engine Selector imeplementation slightly so that unit tests developers could refresh the Singleton when needed, even though that is not needed in normal Dynamo use case. The previous once in a life time initialization was too strict for unit test case especially when multiple Python tests were running across the tests binaries.

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@DynamoDS/dynamo 

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/Libraries/PythonNodeModels/PythonEngineSelector .cs [hunk 1]
```diff
@@ -33,18 +33,26 @@ internal sealed class PythonEngineSelector
         internal static PythonEngineSelector Instance { get { return lazy.Value; } }
 
         // TODO: The following fields might be removed after dynamic loading applied
-        internal static bool IsIronPythonEnabled = false;
-        internal static string IronPythonEvaluatorClass = "IronPythonEvaluator";
-        internal static string IronPythonEvaluationMethod = "EvaluateIronPythonScript";
+        internal bool IsIronPythonEnabled = false;
```
Comment (QilongTang) [inline src/Libraries/PythonNodeModels/PythonEngineSelector .cs:36]: These properties do not need to be static but could rather connected with the Singleton itself. So I updated these

File: src/Libraries/PythonNodeModels/PythonEngineSelector .cs [hunk 2]
```diff
@@ -33,18 +33,26 @@ internal sealed class PythonEngineSelector
...
         /// </summary>
         private PythonEngineSelector()
+        {
+            ScanPythonEngines();
+        }
+
+        /// <summary>
+        /// Scan loaded Python engines
+        /// </summary>
+        internal void ScanPythonEngines()
```
Comment (QilongTang) [inline src/Libraries/PythonNodeModels/PythonEngineSelector .cs:55]: Split the initialization logic out so we can call it in unit test case, when needed, although we rarely need it

File: test/DynamoCoreWpfTests/ViewExtensions/PythonMigrationViewExtensionTests.cs [hunk 3]
```diff
@@ -54,6 +54,7 @@ public void WillDisplayDialogWhenOpeningGraphWithIronPythonNodes()
         [Test]
         public void WillLogNotificationWhenAddingAnIronPythonNode()
         {
+            DebugModes.LoadDebugModesStatusFromConfig(Path.Combine(GetTestDirectory(ExecutingDirectory), "DynamoCoreWpfTests", "python2ObsoleteMode.config"));
```
Comment (QilongTang) [inline test/DynamoCoreWpfTests/ViewExtensions/PythonMigrationViewExtensionTests.cs:57]: Add this so when running this test individually, it also pass..

File: test/DynamoCoreWpfTests/ViewExtensions/PythonMigrationViewExtensionTests.cs [hunk 4]
```diff
@@ -221,9 +221,8 @@ public void CanDetectIronPythonNodesInGraph()
             Open(@"core\python\python.dyn");
 
             var pythonMigration = extensionManager.ViewExtensions
-                .Where(x => x.Name == viewExtension.Name)
-                .Select(x=>x)
-                .First() as PythonMigrationViewExtension;
+                .FirstOrDefault(x => x.Name == viewExtension.Name)
+                as PythonMigrationViewExtension;
```
Comment (QilongTang) [inline test/DynamoCoreWpfTests/ViewExtensions/PythonMigrationViewExtensionTests.cs:225]: Removed some dup code

File: test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs [hunk 5]
```diff
@@ -30,6 +28,9 @@ protected override void GetLibrariesToPreload(List<string> libraries)
         {
             libraries.Add("DSCPython.dll");
             libraries.Add("DSIronPython.dll");
+            // Make sure Python Engine Selector Singleton has all the info up-to-date.
+            // This is not needed for Dynamo in normal use case but useful in unit test case.
+            PythonEngineSelector.Instance.ScanPythonEngines();
```
Comment (mjkkirschner) [inline test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs:33]: will this need to be done when a user installs the ironPython package?

Comment (QilongTang) [reply test/DynamoCoreWpfTests/PythonNodeCustomizationTests.cs:33]: @mjkkirschner Thinking about it.. Yes, regardless of dynamic loading or not, seems need to be called.

**Comment by QilongTang** [comment]
@DynamoDS/dynamo I checked that all unit tests now pass. If you have any comments let me I will still address, I am merging now to make sure bug bash today will be unblocked

**Pull request merged**